### PR TITLE
Readds medium screens

### PR DESCRIPTION
**Pull request by PKPenguin321**
:cl:
add: Integrated circuit medium screens have been readded. They are now called large screens. They now only work from your hands or on the ground when you're standing on top of them (NOT from pockets, lockers, backpacks, etc).
/:cl:

Why:

Circuits may have deserved many of the nerfs in https://github.com/tgstation/tgstation/pull/39376, but the removal of medium screens was absolutely not one of them. It is one of the only meaningful ways to send feedback to the user, and fills the niche of essentially being a `to_chat()` that the TTS circuit doesn't do (TTS announces to EVERYBODY within a screen's range, medium screens only announce to the user and anybody adjacent).

The main reason for removal was "anti-deaf" circuits. The usefulness of those is only questionably useful at best to begin with, as it is very situational (requires you to have printed and to be carrying the circuit with you and then go deaf and then get into a conversation where the other person is unaware that you are deaf). I don't believe this was even close to enough of a reason to remove one of the only meaningful ways of giving user feedback with circuits.

### CODE REVIEW

**Comment by optimumtact** [comment]
no thanks :-1:

**Comment by BlueNothing** [comment]
Are there any particular grounds on which you're objecting?

**Comment by optimumtact** [comment]
yeah, I don't like circuits

**Comment by PKPenguin321** [comment]
Would a "whisper" circuit be okay, where it speaks (requiring you to not be deaf) but only to adjacent tiles?

**Comment by Garen7** [comment]
I didn't elaborate on this in the nerf PR because I was busy, the reason why I removed screens was because it could be added in the left over space of any circuit and still benefit the player by ignoring hearing loss. It could then be put into what ever circuit you're cloning and just always be there similar to floodlights. I know one person had a multi use circuit that just threw in circuits that were strong by themselves and the person would implant it in their chest so they could have an implanted floodlight, ghetto storage implant, smoke machine, and a screen as a way to ignore hearing loss. I removed screens because I knew they could've taken the place of floodlights once that got removed as well as to nerf that multiuse implant.

I intend to work on the ui of circuits until cloning is no longer needed to make any kind of complex circuit and then remove cloning. Once I remove cloning people putting random multiuse circuits into their cloning save file no longer becomes an issue so I'm going to PR to return screens when I remove cloning. I would like to remove cloning now but that would mean that anyone doing circuits could only make basic circuits like chatspammers and other useless, uninteresting things.

Also I said in the nerf PR that if anyone missed the short range of medium screens then they should PR a whisper circuits since that would be plenty fine balance wise.

**Comment by deathride58** [comment]
Automated speech-to-text circuits are actually a pretty smart use of circuitry, and are only reasonably useful in very specific edge cases. If they're really, truly that big of an issue, then make it so that speech/audio is a technically separate data type from text, so that you have to manually "convert" microphone input into text with a module that runs a filter over a text to simulate the usual speech-to-text inaccuracies found in most speech-to-text implementations in the real world, and sets the data type to text. Removing small emergent gameplay just because of a singular edge case that requires a very specific, rare condition in order to even be useful at all is just ridiculous.

**Comment by BlueNothing** [comment]
I'd be okay with a whisper circuit or medium screens returning or both. From what I understand, though, large screens, as things are, would be seen as a bridge too far.

**Comment by GuyonBroadway** [comment]
I mean, making an anti-deaf circuit is more effort than getting a pillbottle of the deafness healing chem from chemistry and has the same effect.

**Comment by BlueNothing** [comment]
The screens PK wants to restore here have tremendous utility to circuit users and developers of all levels, from someone who just picked the thing up and wants a guide on how to use it, to someone doing cutting-edge work. People say that the problems are either anti-deaf circuits (which are rather situational, and *still possible*) or their use in pretty much every kind of circuit (which is a function of how incredibly useful they were for most tasks). Neither of these appear solid arguments for their removal in the first place.

**Comment by Jalleo** [comment]
Isnt the proper solution preventing the implabt circuit possibility?
Seriously removing a screen for a situational benefit where you can shout I AM DEAF and then using me is still going to be the main way to talk.

Both these PRs are bad and focus in on a situation of powergaming. Screens should only work if they are visible as in on a hand not in storage and not inside someones stomach.

That isnt for balance its for giving people a proper thoughtful design basis.

**Comment by PKPenguin321** [comment]
>Seriously removing a screen for a situational benefit where you can shout I AM DEAF and then using me is still going to be the main way to talk.

I don't really understand, could you rephrase?

**Comment by BlueNothing** [comment]
They're saying:
>Normal day on station.
>Bomb comes up.
>MFW deaf.
>Me "fans (or covers) ears."
>Other person: Me "nods. 'Hello, there.'"
I think that's the situation being described. I don't see that particular solution happen much, nor a lot of miming, but typically, I think standard operating procedure is to just get inacusiate from chemistry and call it good.

**Comment by optimumtact** [comment]
Make the screens not function when the circuit is inside something else if they don't already and this can be merged.

**Comment by PKPenguin321** [comment]
@optimumtact done and it's tested and works

File: code/modules/integrated_electronics/subtypes/output.dm [hunk 1]
```diff
@@ -34,6 +34,34 @@
...
+	power_draw_per_use = 20
+
+/obj/item/integrated_circuit/output/screen/large/do_work()
+	..()
+
+	var/in_hands = FALSE//this whole block just returns if the assembly is neither in a mobs hands or on the ground
+	if(isliving(assembly.loc))
+		in_hands = TRUE
+		var/mob/living/H = assembly.loc
+		if(H.get_active_held_item() != assembly && H.get_inactive_held_item() != assembly)
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/integrated_electronics/subtypes/output.dm]: can you explain the point of `in_hands` when this line exists?

**Review by ExcessiveUseOfCobblestone — CHANGES_REQUESTED** [review]

File: code/modules/integrated_electronics/subtypes/output.dm (hunk 1 above)
Comment (PKPenguin321) [reply code/modules/integrated_electronics/subtypes/output.dm]: sure! if you look further down at line 53, i need to check a second time whether or not the assembly is in hands. if i don't, then at line 53 it will return the whole proc even if it passes this check here at line 51. in_hands is set to TRUE here so that if this part succeeds, then the check at line 53 will also succeed. you'll notice it's set to TRUE before the check for whether or not it actually is in hands or not happens: this is because if the check fails, then the proc returns anyways despite that var being TRUE.
tl;dr code logic reasons

Comment (PKPenguin321) [reply code/modules/integrated_electronics/subtypes/output.dm]: basically to make it simpler, if `in_hands` didn't exist then even if the object actually IS in somebody's hands, the code would still return at line 53. so we need it to make that not happen

Comment (optimumtact) [reply code/modules/integrated_electronics/subtypes/output.dm]: it's probably just because you never actually use it, so it looks weird to devs

stylistically you could do
```
var/in_hands = (H.get_active_held_item() = assembly || H.get_inactive_held_item() = assembly)
if(!in_hands)
   return
if(!isturf(assembly.loc) && !in_hands)
   return
```
but then it's obvious that you don't need the first check?

I'm confused

**Comment by PKPenguin321** [comment]
@optimumtact looks better now?

**Comment by ExcessiveUseOfCobblestone** [comment]
I'm not sure why you need the var at all

**Comment by ExcessiveUseOfCobblestone** [comment]
```
	if(isliving(assembly.loc))
		var/mob/living/H = assembly.loc
		if(H.get_active_held_item() != assembly && H.get_inactive_held_item() != assembly)
			return
	if(!isturf(assembly.loc))
                return
```

is the same as current but without the var

**Comment by PKPenguin321** [comment]
No, because then if it is in your hands it will return when it reaches `if(!isturf(assembly.loc))` despite the fact that it's supposed to work from your hands

**Comment by ExcessiveUseOfCobblestone** [comment]
fixed

**Comment by PKPenguin321** [comment]
yeah that works

File: code/modules/integrated_electronics/subtypes/output.dm [hunk 2]
```diff
@@ -34,6 +34,33 @@
...
+
+/obj/item/integrated_circuit/output/screen/large/do_work()
+	..()
+
+	if(isliving(assembly.loc))//this whole block just returns if the assembly is neither in a mobs hands or on the ground
+		var/mob/living/H = assembly.loc
+		if(H.get_active_held_item() != assembly && H.get_inactive_held_item() != assembly)
+			return
+	else
+		if(!isturf(assembly.loc))
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/integrated_electronics/subtypes/output.dm]: 51/52 can be one line but it's okay unless contributing.md or maintainer or both say otherwise

**Review by ExcessiveUseOfCobblestone — APPROVED** [review]

File: code/modules/integrated_electronics/subtypes/output.dm (hunk 2 above)
Comment (ExcessiveUseOfCobblestone) [reply code/modules/integrated_electronics/subtypes/output.dm]: github is weird but I gave you the cute irrelevant check
